src-self-hosted/dep_tokenizer.zig

```diff
@@ -0,0 +1,1131 @@
+        return Error.InvalidInput;
+    }
+
+    fn errorPosition(self: *Tokenizer, position: usize, bytes: []const u8, comptime fmt: []const u8, args: ...) Error {
+        var buffer = try std.Buffer.initSize(&self.arena.allocator, 0);
+        std.fmt.format(&buffer, anyerror, std.Buffer.append, fmt, args) catch {};
+        try buffer.append(" '");
+        var out = makeOutput(std.Buffer.append, &buffer);
+        try printCharValues(&out, bytes);
+        try buffer.append("'");
+        std.fmt.format(&buffer, anyerror, std.Buffer.append, " at position {}", position - (bytes.len - 1)) catch {};
+        self.error_text = buffer.toSlice();
+        return Error.InvalidInput;
+    }
+
+    fn errorIllegalChar(self: *Tokenizer, position: usize, char: u8, comptime fmt: []const u8, args: ...) Error {
+        var buffer = try std.Buffer.initSize(&self.arena.allocator, 0);
+        try buffer.append("illegal char ");
+        var out = makeOutput(std.Buffer.append, &buffer);
+        try printUnderstandableChar(&out, char);
+        std.fmt.format(&buffer, anyerror, std.Buffer.append, " at position {}", position) catch {};
+        if (fmt.len != 0) std.fmt.format(&buffer, anyerror, std.Buffer.append, ": " ++ fmt, args) catch {};
+        self.error_text = buffer.toSlice();
+        return Error.InvalidInput;
+    }
+
+    const Error = error{
+        OutOfMemory,
+        InvalidInput,
+    };
+
+    const State = union(enum) {
+        lhs: void,
+        target: std.Buffer,
+        target_reverse_solidus: std.Buffer,
+        target_dollar_sign: std.Buffer,
+        target_colon: std.Buffer,
+        target_colon_reverse_solidus: std.Buffer,
+        rhs: void,
+        rhs_continuation: void,
+        rhs_continuation_linefeed: void,
+        prereq_quote: std.Buffer,
+        prereq: std.Buffer,
+        prereq_continuation: std.Buffer,
+        prereq_continuation_linefeed: std.Buffer,
+    };
+
+    const Token = struct {
+        id: ID,
+        bytes: []const u8,
+
+        const ID = enum {
+            target,
+            prereq,
+        };
+    };
+};
+
+// stage1 compiler support
+var stage2_da = std.heap.DirectAllocator.init();
+
+export fn stage2_DepTokenizer_init(input: [*]const u8, len: usize) stage2_DepTokenizer {
+    const t = stage2_da.allocator.create(Tokenizer) catch unreachable;
+    t.* = Tokenizer.init(&stage2_da.allocator, input[0..len]);
+    return stage2_DepTokenizer{
+        .handle = t,
+    };
+}
+
+export fn stage2_DepTokenizer_deinit(self: *stage2_DepTokenizer) void {
+    self.handle.deinit();
+}
+
+export fn stage2_DepTokenizer_next(self: *stage2_DepTokenizer) stage2_DepNextResult {
+    const otoken = self.handle.next() catch {
+        return stage2_DepNextResult{
+            .ent = 0,
+            .textz = (std.Buffer.init(&self.handle.arena.allocator, self.handle.error_text) catch unreachable).toSlice().ptr,
```

Comment:
`catch unreachable` is not justified here

Reply:
please clarify; is that to mean instead use @panic() or to mean coding around it, return an error with some kind of hardcoded text?

Reply:
`@panic` is fine - that's what stage1 currently does on OOM. But `catch unreachable` invokes undefined behavior in release builds. We definitely don't want that since we can't guarantee that memory allocation will always succeed.